libs/utils/trace_analysis.py

```diff
@@ -1009,52 +888,107 @@ def getClusterFrequencyResidency(self, cluster):
-        for c in range(1, len(activations.columns)):
-            cluster_active |= activations[activations.columns[c]].astype(int)
-
-        start = []
-        end = []
-        active = False
-        for idx, val in cluster_active.iteritems():
-            if val == 1 and not active:
-                start.append(idx)
-                active = True
-            elif val == 0 and active:
-                end.append(idx)
-                active = False
-        if len(start) > len(end):
-            end.append(self.x_max)
-
-        tasks_time_intervals = pd.DataFrame({
-            "start" : start,
-            "end"   : end
-        })
+        for c in activations.columns[1:]:
+            cluster_active |= activations[c].astype(int)
+
+        return cluster_active
+
+    def _getCPUActiveSignal(self, cpu):
+        """
+        Build a square wave representing the active (i.e. non-idle) CPU time,
+        i.e.:
+            cpu_active[t] = 1 if there is a task running in the CPU at time t
+            cpu_active[t] = 0 otherwise
+
+        :param cpu: CPU ID
+        :type cpu: int
+        """
+        return self._getClusterActiveSignal([cpu])
+
+    def _integrate_square_wave(self, sq_wave):
+        """
+        Compute the integral of a square wave time series.
+
+        :param sq_wave: square wave assuming only 1.0 and 0.0 values
+        :type sq_wave: :mod:`pandas.Series`
+        """
+        sq_wave.iloc[-1] = 0.0
+        # Compact signal to obtain only 1-0-1-0 sequences
+        comp_sig = sq_wave.loc[sq_wave.shift() != sq_wave]
+        # First value for computing the difference must be a 1
+        if comp_sig.iloc[0] == 0.0:
+            return sum(comp_sig.iloc[2::2].index - comp_sig.iloc[1:-1:2].index)
+        else:
+            return sum(comp_sig.iloc[1::2].index - comp_sig.iloc[:-1:2].index)
+
+    def getClusterFrequencyResidency(self, cluster):
+        """
+        Get a DataFrame with per cluster frequency residency, i.e. amount of
+        time spent at a given frequency in each cluster.
 
-        # Compute the non-IDLE time spent at a certain frequency
-        time_intervals = []
-        frequencies = []
-        row_iterator = cluster_freqs.iterrows()
-        # Take first item from row_iterator
-        t_prev, f_prev = row_iterator.next()
-        for t, f in row_iterator:
-            nonidle_time = self._computeNonIdleTime(t_prev,
-                                                    t,
-                                                    tasks_time_intervals)
-            freq = f_prev['frequency']
-            time_intervals.append(nonidle_time)
-            frequencies.append(freq)
-            t_prev = t
-            f_prev = f
-
-        active_time = pd.DataFrame({
-            'ACTIVE Time' : time_intervals,
-            'Frequency [MHz]' : [f/1000 for f in frequencies]
+        :param cluster: list of CPU IDs belonging to the cluster
+        :type cluster: list(int)
+        """
+
```

Comment:
What about allowing cluster to be just a string if we want just one cluster?

In this case, this should make the trick for the following code:
```python
if not isinstance(cluster, list):
    cluster = [cluster]
```